core/ws_comment_events.js

```diff
@@ -0,0 +1,410 @@
+  this.xml = Blockly.Xml.textToDom('<xml>' + json['xml'] + '</xml>').firstChild;
+};
+
+/**
+ * Run a creation event.
+ * @param {boolean} forward True if run forward, false if run backward (undo).
+ */
+Blockly.Events.CommentCreate.prototype.run = function(forward) {
+  var workspace = this.getEventWorkspace_();
+  if (forward) {
+    var xml = goog.dom.createDom('xml');
+    xml.appendChild(this.xml);
+    Blockly.Xml.domToWorkspace(xml, workspace);
+  } else {
+    var comment = workspace.getCommentById(this.commentId);
+    if (comment) {
+      comment.dispose(false, false);
+    } else {
+      // Only complain about root-level block.
+      console.warn("Can't uncreate non-existent comment: " + this.commentId);
+    }
+  }
+};
+
+/**
+ * Class for a comment deletion event.
+ * @param {Blockly.WorkspaceComment} comment The deleted comment.
+ *     Null for a blank event.
+ * @extends {Blockly.Events.CommentBase}
+ * @constructor
+ */
+Blockly.Events.CommentDelete = function(comment) {
+  if (!comment) {
+    return;  // Blank event to be populated by fromJson.
+  }
+  Blockly.Events.CommentDelete.superClass_.constructor.call(this, comment);
+
+  this.xml = comment.toXmlWithXY();
+};
+goog.inherits(Blockly.Events.CommentDelete, Blockly.Events.CommentBase);
+
+/**
+ * Type of this event.
+ * @type {string}
+ */
+Blockly.Events.CommentDelete.prototype.type = Blockly.Events.COMMENT_DELETE;
+
+/**
+ * Encode the event as JSON.
+ * TODO (#1266): "Full" and "minimal" serialization.
+ * @return {!Object} JSON representation.
+ */
+Blockly.Events.CommentDelete.prototype.toJson = function() {
+  var json = Blockly.Events.CommentDelete.superClass_.toJson.call(this);
+  return json;
+};
+
+/**
+ * Decode the JSON event.
+ * @param {!Object} json JSON representation.
+ */
+Blockly.Events.CommentDelete.prototype.fromJson = function(json) {
+  Blockly.Events.CommentDelete.superClass_.fromJson.call(this, json);
+};
+
+/**
+ * Run a creation event.
+ * @param {boolean} forward True if run forward, false if run backward (undo).
+ */
+Blockly.Events.CommentDelete.prototype.run = function(forward) {
+  var workspace = this.getEventWorkspace_();
+  if (forward) {
+    var comment = workspace.getCommentById(this.commentId);
+    if (comment) {
+      comment.dispose(false, false);
+    } else {
+      // Only complain about root-level block.
+      console.warn("Can't uncreate non-existent comment: " + this.commentId);
+    }
+  } else {
+    var xml = goog.dom.createDom('xml');
+    xml.appendChild(this.xml);
+    Blockly.Xml.domToWorkspace(xml, workspace);
+  }
+};
+
+/**
+ * Class for a comment move event.  Created before the move.
+ * @param {Blockly.WorkspaceComment} comment The comment that is being moved.
+ *     Null for a blank event.
+ * @extends {Blockly.Events.CommentBase}
+ * @constructor
+ */
+Blockly.Events.CommentMove = function(comment) {
+  if (!comment) {
+    return;  // Blank event to be populated by fromJson.
```

Comment:
to be populated by recordNew?

Reply:
actually maybe nvm...

Reply:
Yeah, if there's no comment it'll be populated by `fromJson`.  If there's a comment we record the old coordinate but still need to call `recordNew` later.